completion/available/pip.completion.bash

```diff
@@ -2,9 +2,36 @@
 # Of course, you should first install the pip, say on Debian:
 # sudo apt-get install python-pip
 # sudo apt-get install python3-pip
-# If the pip package is installed within virtual environments, say, python managed by pyenv,
-# you should first initilization the corresponding environment.
-# So that the pip/pip3 is in system's path.
-if command -v pip >/dev/null; then
-  eval "$(pip completion --bash)"
+
+# If the pip package is installed within virtual environments,
+# you should first install pip for the corresponding environment.
+
+# Fix pip completion for running it within/outside of pyenv/virtualenv/venv/conda environment.
+_regex="(^|;[ ]*)_pip_completion_hook([ ]*;|$)"
+
+_pip_completion_hook() {
+  local _pip
+  # For pip resides within the pyenv/virtualenv/venv/conda environments:
+  if [ -n "$VIRTUAL_ENV" ] && [ -x "$VIRTUAL_ENV/bin/pip" ]; then
+    _pip="$VIRTUAL_ENV/bin/pip"
+  elif [ -n "$CONDA_PREFIX" ] && [ -x "$CONDA_PREFIX/bin/pip" ]; then
+    _pip="$CONDA_PREFIX/bin/pip"
+  # For pip resides outside of a virtual environment:
+  elif [ -x /usr/bin/pip ]; then
+    _pip=/usr/bin/pip
+  fi
+  if [ -n "$_pip" ]; then
+    eval "$($_pip completion --bash)"
```

Comment:
I understand what you're trying to do here, but doesn't this add significant overhead to each prompt? You're basically reevaluating the pip completion on every prompt, right? Not sure that's a good idea, as it might slow down things significantly.

Reply:
The problem is that pip maybe existed/installed for each virtualenv. You cannot expect the system's pip is the same version as the currently activated virtualenv.

In other hand, the suggested usage of pip with virutalenv is set the following config option:

require-virtualenv = true

to prevent the interactions between the packaged among different virtualenvs.

This also will result in the pip installed for each virtualenv separately.

If you activated a virutalenv, you should use the completion given by pip installed for this  virutalenv.  I cannot find more efficient method for solving this problem.

Based on my experiences, this will add about 0.2s delay on the appearance of the new prompt.  You can try to feel it.

Regards

Reply:
I'm sorry, but I don't agree with this approach at all. The existing version (which uses `eval`) continues to work even if there are changes in `pip`. With your new version, you have added a maintenance burden to Bash-it, where the completion code needs to be updated if there are changes in `pip`'s completion code.

I don't see this as an improvement over the current version in Bash-it - I'd prefer to keep things like they are in this case.

Reply:
The eval based method has the following problem when working with multiple virtualenvs:

Say, I haven't pip installed system-wide, and I have the pip installed only within the virtualenv.  In this case, when you activate a virtualenv, the pip completion will failed to take effect till you add the eval relative command to the PROMPT_COMMAND.

Regards